### ISSUE TITLE
scaling down deployment not considering affinity

### ISSUE THREAD

**Issue by barucoh** [kubernetes/kubernetes#89922]
**What happened**:
A deployment with 3 replicas with anti-affinity rule to spread across 2 AZs scaled down to 2 replicas in only 1 AZ

**What you expected to happen**:
Scale down operations, like scale up operations, should consider anti-affinity rules or `SelectorSpreadPriority`

**How to reproduce it (as minimally and precisely as possible)**:
K8s cluster across 2 AZs, single deployment, 3 replicas, `preferred` anti-affinity rule with a weight of 100 to spread the pods across AZs.
Pods are spread 2-1, make sure the AZ with the 1 pod has other extra pods to "simulate higher usage".
scale down the deployment.

**Anything else we need to know?**:
I'm using Kubernetes `1.14.9` on EKS

**Environment**:
- Kubernetes version (use `kubectl version`):
`v1.14.9-eks-502bfb`
- Cloud provider or hardware configuration:
AWS EKS 1.14

**Comment by ahg-g** [comment]
the scheduler isn't involved in deployment scale down, and so it can't help here.

a descheduler could help here, but only after the fact.

**Comment by barucoh** [comment]
> the scheduler isn't involved in deployment scale down, and so it can't help here.
I probably put the incorrect sig of scheduling.

> a descheduler could help here, but only after the fact.
If you mean the [Descheduler for Kubernetes](https://github.com/kubernetes-sigs/descheduler) then I'm afraid it doesn't seem to solve the issue since it's also not involved when a scale down operation is taking place, plus it only looks at `required` anti-affinity rules.

But bottom line it doesn't really matter which component is required for this to happen, but I think this is an important feature.
You can have the kube-scheduler schedule 10 pods across 3 different AZs as 3-3-4 best case, and HPA scale down 6 of them leaving 3-0-0 resulting in a degraded state from an Availability Zone HA point of view.

**Comment by Huang-Wei** [comment]
@barucoh As @ahg-g pointed out, scaling down a workload API is implemented in controller-manager, but unfortunately, controller-manager doesn't have the latest view of cluster utilization, hence may give non-optimal result.

I'd say this is a known limitation for a while, but we're open to proposals improving current situation.

**Comment by barucoh** [comment]
I dug a bit more and noticed my issue is basically this https://github.com/kubernetes/kubernetes/issues/4301

I feel like this can be closed, but the other issue is opened for over 5 years now...
I'll try to think of a proposal for this

**Comment by Huang-Wei** [comment]
We can keep this open for now, and close this until we have a promising proposal.

**Comment by gongx** [comment]
I am very interested in this issue. I read some of the source codes about the replica-set. It looks like during the scaling down,  we are using ActivePodsWithRanks to sort and pick deletion candidates. Does this the right place to check? If it is, maybe we can customize the rank value based on the affinity or anti-affinity to solve it.

**Comment by Huang-Wei** [comment]
It's pretty challenging to do the ranking in controller-manager side as you have to know the whole cluster's status to do the right decision. That's out of controll-manager's knowledge.

**Comment by barucoh** [comment]
Personally, one of the requirements to be truly production ready is this.
If the cluster cannot guarantee we have availability across hosts/zones/regions at any given time and only during deployment time, trusting the system becomes really hard.

@gongx In my case, as I wanted to enforce availability across zones, I was forced to have a number of deployments, 1 per AZ with an HPA each. :cry:

**Comment by gongx** [comment]
@barucoh  
If we set a 'strict' requirement for affinity/anti-affinity, the scaling down deployment will definitely honor the setting. 
If we set a 'prefer' requirement for affinity/anti-affinity,  it means that we are trying the best to honor the setting, and even the deployment, before scaling down, can not be guaranteed. So, after scaling down, the deployment does not fully honor the setting should be fine as well.

@Huang-Wei I totally agree with you that the controller manager does not know the whole cluster's status and can not make the best decision. It will be too complex to add a similar service as descheduler in controller-manager. But given this is the 'prefer' case, if the controller-manager can calculate the ranking just based on the current deployment/replica-set status and instead of using hard-code way, It may make a better decision.
